### ISSUE TITLE
yaml RepresenterError in TestClientRecorder for CIM objects

### ISSUE THREAD

**Issue by KSchopmeyer** [pywbem/pywbem#588]
When executing one of the inovoke methods in run_cimoperations.py that returns an exception
we get the following issue from the recorder.
```
======================================================================
ERROR: test_all (__main__.InvokeMethod)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/kschopmeyer/pywbem/githubpywbem/pywbem/pywbem/cim_operations.py", line 4808, in InvokeMethod
    result = self._methodcall(MethodName, obj, Params, **params)
  File "/home/kschopmeyer/pywbem/githubpywbem/pywbem/pywbem/cim_operations.py", line 1112, in _methodcall
    raise CIMError(code, tt[0][1]['DESCRIPTION'])
pywbem.exceptions.CIMError: 16: CIM_ERR_METHOD_NOT_AVAILABLE: FooMethod

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "testsuite/run_cim_operations.py", line 1939, in test_all
    Ref=name)
  File "testsuite/run_cim_operations.py", line 126, in cimcall
    result = fn(*pargs, **kwargs)
  File "/home/kschopmeyer/pywbem/githubpywbem/pywbem/pywbem/cim_operations.py", line 4834, in InvokeMethod
    self.operation_recorder.record_staged()
  File "/home/kschopmeyer/pywbem/githubpywbem/pywbem/pywbem/_recorder.py", line 300, in record_staged
    self.record(pwargs, pwresult, httpreq, httpresp)
  File "/home/kschopmeyer/pywbem/githubpywbem/pywbem/pywbem/_recorder.py", line 453, in record
    default_flow_style=False)
  File "/home/kschopmeyer/.virtualenvs/pywbem3/lib/python3.4/site-packages/yaml/__init__.py", line 216, in safe_dump
    return dump_all([data], stream, Dumper=SafeDumper, **kwds)
  File "/home/kschopmeyer/.virtualenvs/pywbem3/lib/python3.4/site-packages/yaml/__init__.py", line 188, in dump_all
    dumper.represent(data)
  File "/home/kschopmeyer/.virtualenvs/pywbem3/lib/python3.4/site-packages/yaml/representer.py", line 26, in represent
    node = self.represent_data(data)
  File "/home/kschopmeyer/.virtualenvs/pywbem3/lib/python3.4/site-packages/yaml/representer.py", line 47, in represent_data
    node = self.yaml_representers[data_types[0]](self, data)
  File "/home/kschopmeyer/.virtualenvs/pywbem3/lib/python3.4/site-packages/yaml/representer.py", line 195, in represent_list
    return self.represent_sequence('tag:yaml.org,2002:seq', data)
  File "/home/kschopmeyer/.virtualenvs/pywbem3/lib/python3.4/site-packages/yaml/representer.py", line 91, in represent_sequence
    node_item = self.represent_data(item)
  File "/home/kschopmeyer/.virtualenvs/pywbem3/lib/python3.4/site-packages/yaml/representer.py", line 47, in represent_data
    node = self.yaml_representers[data_types[0]](self, data)
  File "/home/kschopmeyer/pywbem/githubpywbem/pywbem/pywbem/_recorder.py", line 83, in <lambda>
    _represent_ordereddict(dumper, u'tag:yaml.org,2002:map', value))
  File "/home/kschopmeyer/pywbem/githubpywbem/pywbem/pywbem/_recorder.py", line 64, in _represent_ordereddict
    node_value = dump.represent_data(item_value)
  File "/home/kschopmeyer/.virtualenvs/pywbem3/lib/python3.4/site-packages/yaml/representer.py", line 47, in represent_data
    node = self.yaml_representers[data_types[0]](self, data)
  File "/home/kschopmeyer/pywbem/githubpywbem/pywbem/pywbem/_recorder.py", line 83, in <lambda>
    _represent_ordereddict(dumper, u'tag:yaml.org,2002:map', value))
  File "/home/kschopmeyer/pywbem/githubpywbem/pywbem/pywbem/_recorder.py", line 64, in _represent_ordereddict
    node_value = dump.represent_data(item_value)
  File "/home/kschopmeyer/.virtualenvs/pywbem3/lib/python3.4/site-packages/yaml/representer.py", line 47, in represent_data
    node = self.yaml_representers[data_types[0]](self, data)
  File "/home/kschopmeyer/pywbem/githubpywbem/pywbem/pywbem/_recorder.py", line 83, in <lambda>
    _represent_ordereddict(dumper, u'tag:yaml.org,2002:map', value))
  File "/home/kschopmeyer/pywbem/githubpywbem/pywbem/pywbem/_recorder.py", line 64, in _represent_ordereddict
    node_value = dump.represent_data(item_value)
  File "/home/kschopmeyer/.virtualenvs/pywbem3/lib/python3.4/site-packages/yaml/representer.py", line 57, in represent_data
    node = self.yaml_representers[None](self, data)
  File "/home/kschopmeyer/.virtualenvs/pywbem3/lib/python3.4/site-packages/yaml/representer.py", line 227, in represent_undefined
    raise RepresenterError("cannot represent an object: %s" % data)
yaml.representer.RepresenterError: cannot represent an object: 00000060000000.000000:000
```

**Comment by andy-maier** [comment]
PyYAML is trying to handle a `CIMDateTime` object and does not find a representer for it.

Looking at the PyYAML code, it seems that its representer goes by the type of the Python object, not considering inheritance:

Code from `yaml.BaseRepresenter.represent_data()`:
```
        data_types = type(data).__mro__
        if data_types[0] in self.yaml_representers:
            node = self.yaml_representers[data_types[0]](self, data)
        else:
            for data_type in data_types:
                if data_type in self.yaml_multi_representers:
                    node = self.yaml_multi_representers[data_type](self, data)
                    break
            else:
                if None in self.yaml_multi_representers:
                    node = self.yaml_multi_representers[None](self, data)
                elif None in self.yaml_representers:
                    node = self.yaml_representers[None](self, data)
                else:
                    node = ScalarNode(None, str(data))
```
The representers are all in `self.yaml_representers`, which is used only for the exact Python type of the object (not considering inheritance). The representers in `self.yaml_multi_representers` would also be used for the Python types in the MRO list (considering inheritance), but it does not have any representers.

Bottom line: It seems for PyYAML to be able to dump any of the CIM object types of Pywbem, we need representer functions for each of them.

**Comment by andy-maier** [comment]
PR #613 improves the diagnostics for the unknown data type case.

**Comment by andy-maier** [comment]
A new update to PR #613 fixes a new `AttributeError` that was raised when invoking `run_cim_operations.py` with a non-existing file url (e.g. `blah`).

The original error reported in this issue still needs to be reproduced with PR #613.

**Comment by KSchopmeyer** [comment]
Since the original problem came because there is no representer for CIMDateTime, this should still have caused a problem.  Karl will retest.

**Comment by KSchopmeyer** [comment]
Note that the representer is still not working.  This is the output.
```
======================================================================
ERROR: test_all_param_types (__main__.InvokeMethod)
Call invoke method with all possible parameter types
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./run_cim_operations.py", line 2316, in test_all_param_types
    Ref=name)
  File "./run_cim_operations.py", line 153, in cimcall
    result = fn(*pargs, **kwargs)
  File "/home/kschopmeyer/.virtualenvs/pywbem27/local/lib/python2.7/site-packages/pywbem-0.10.0.dev0-py2.7.egg/pywbem/cim_operations.py", line 6892, in InvokeMethod
    self.operation_recorder.record_staged()
  File "/home/kschopmeyer/.virtualenvs/pywbem27/local/lib/python2.7/site-packages/pywbem-0.10.0.dev0-py2.7.egg/pywbem/_recorder.py", line 321, in record_staged
    self.record(pwargs, pwresult, httpreq, httpresp)
  File "/home/kschopmeyer/.virtualenvs/pywbem27/local/lib/python2.7/site-packages/pywbem-0.10.0.dev0-py2.7.egg/pywbem/_recorder.py", line 481, in record
    default_flow_style=False)
  File "/home/kschopmeyer/.virtualenvs/pywbem27/local/lib/python2.7/site-packages/yaml/__init__.py", line 218, in safe_dump
    return dump_all([data], stream, Dumper=SafeDumper, **kwds)
  File "/home/kschopmeyer/.virtualenvs/pywbem27/local/lib/python2.7/site-packages/yaml/__init__.py", line 190, in dump_all
    dumper.represent(data)
  File "/home/kschopmeyer/.virtualenvs/pywbem27/local/lib/python2.7/site-packages/yaml/representer.py", line 28, in represent
    node = self.represent_data(data)
  File "/home/kschopmeyer/.virtualenvs/pywbem27/local/lib/python2.7/site-packages/yaml/representer.py", line 57, in represent_data
    node = self.yaml_representers[data_types[0]](self, data)
  File "/home/kschopmeyer/.virtualenvs/pywbem27/local/lib/python2.7/site-packages/yaml/representer.py", line 215, in represent_list
    return self.represent_sequence(u'tag:yaml.org,2002:seq', data)
  File "/home/kschopmeyer/.virtualenvs/pywbem27/local/lib/python2.7/site-packages/yaml/representer.py", line 101, in represent_sequence
    node_item = self.represent_data(item)
  File "/home/kschopmeyer/.virtualenvs/pywbem27/local/lib/python2.7/site-packages/yaml/representer.py", line 57, in represent_data
    node = self.yaml_representers[data_types[0]](self, data)
  File "/home/kschopmeyer/.virtualenvs/pywbem27/local/lib/python2.7/site-packages/pywbem-0.10.0.dev0-py2.7.egg/pywbem/_recorder.py", line 83, in <lambda>
    _represent_ordereddict(dumper, u'tag:yaml.org,2002:map', value))
  File "/home/kschopmeyer/.virtualenvs/pywbem27/local/lib/python2.7/site-packages/pywbem-0.10.0.dev0-py2.7.egg/pywbem/_recorder.py", line 64, in _represent_ordereddict
    node_value = dump.represent_data(item_value)
  File "/home/kschopmeyer/.virtualenvs/pywbem27/local/lib/python2.7/site-packages/yaml/representer.py", line 57, in represent_data
    node = self.yaml_representers[data_types[0]](self, data)
  File "/home/kschopmeyer/.virtualenvs/pywbem27/local/lib/python2.7/site-packages/pywbem-0.10.0.dev0-py2.7.egg/pywbem/_recorder.py", line 83, in <lambda>
    _represent_ordereddict(dumper, u'tag:yaml.org,2002:map', value))
  File "/home/kschopmeyer/.virtualenvs/pywbem27/local/lib/python2.7/site-packages/pywbem-0.10.0.dev0-py2.7.egg/pywbem/_recorder.py", line 64, in _represent_ordereddict
    node_value = dump.represent_data(item_value)
  File "/home/kschopmeyer/.virtualenvs/pywbem27/local/lib/python2.7/site-packages/yaml/representer.py", line 57, in represent_data
    node = self.yaml_representers[data_types[0]](self, data)
  File "/home/kschopmeyer/.virtualenvs/pywbem27/local/lib/python2.7/site-packages/pywbem-0.10.0.dev0-py2.7.egg/pywbem/_recorder.py", line 83, in <lambda>
    _represent_ordereddict(dumper, u'tag:yaml.org,2002:map', value))
  File "/home/kschopmeyer/.virtualenvs/pywbem27/local/lib/python2.7/site-packages/pywbem-0.10.0.dev0-py2.7.egg/pywbem/_recorder.py", line 64, in _represent_ordereddict
    node_value = dump.represent_data(item_value)
  File "/home/kschopmeyer/.virtualenvs/pywbem27/local/lib/python2.7/site-packages/yaml/representer.py", line 67, in represent_data
    node = self.yaml_representers[None](self, data)
  File "/home/kschopmeyer/.virtualenvs/pywbem27/local/lib/python2.7/site-packages/yaml/representer.py", line 247, in represent_undefined
    raise RepresenterError("cannot represent an object: %s" % data)
RepresenterError: cannot represent an object: 00000060000000.000000:000
```

**Comment by andy-maier** [comment]
We concluded in the call today that the support for CIM data types needs to be added to the TestClientRecorder, and that this does not need to be done for 0.10.0.

**Comment by andy-maier** [comment]
Update for Karl: Please re-test whether the original issue is still in 0.11.0.

**Comment by andy-maier** [comment]
This seems to be a duplicate of PR #702.

**Comment by andy-maier** [comment]
Fixed by PR #984.
